File: recipes/07_mesh_rpc.ts
/**
 * Recipe 07 — Mesh request/reply (the canonical app RPC pattern).
 *
 * Teaches: device-to-device app messaging over the mesh. A server subscribes to a topic and
 * answers inbound requests via `mesh.reply(replyToken, ...)`; a client calls
 * `mesh.request(node, topic, payload, timeout)` and awaits the reply. This is the building block
 * every CE *app* uses for cross-node features (the node stays primitives-only).
 *
 * Run:  npx tsx recipes/07_mesh_rpc.ts
 *
 * needs: two-nodes (a real request needs a peer running the reply loop). With one node we
 * exercise subscribe + the inbox snapshot, the single-node-testable slice of the contract.
 * Set CE_PEER_NODE_ID to run a live two-node round-trip.
 */
import { CeClient, utf8ToBytes, bytesToUtf8 } from "@ce-net/sdk";

async function main(): Promise<void> {
  const baseUrl = process.env["CE_BASE_URL"] ?? "http://127.0.0.1:8844";
  const token = process.env["CE_API_TOKEN"];
  const ce = token ? new CeClient({ baseUrl, token }) : CeClient.local();

  const topic = "cookbook/ping";

  // Server side: subscribe so inbound requests land in the inbox + stream.
  await ce.mesh.subscribe(topic);
  console.log(`subscribed to '${topic}'`);

  // The reply loop an app would run (drains requests, answers each):
  //
  //   for await (const m of ce.mesh.streamMessages()) {
  //     if (m.topic === topic && m.replyToken != null) {
  //       await ce.mesh.reply(m.replyToken, utf8ToBytes("pong"));
  //     }
  //   }
  //
  // Client side, on a *different* node:
  //
  //   const reply = await ce.mesh.request(serverNodeId, topic, new Uint8Array(), 5000);
  //   console.log(bytesToUtf8(reply)); // "pong"

  // Single-node-testable slice: the inbox snapshot is reachable and returns a (possibly empty) list.
  const inbox = await ce.mesh.messages();
  console.log(`inbox snapshot: ${inbox.length} message(s)`);

  const peer = process.env["CE_PEER_NODE_ID"];
  if (peer) {
    try {
      const reply = await ce.mesh.request(peer, topic, utf8ToBytes(""), 5000);
      console.log(`round-trip reply: ${bytesToUtf8(reply)}`);
    } catch (e) {
      console.log(`peer round-trip skipped: ${e instanceof Error ? e.message : String(e)}`);
    }
  } else {
    console.log("(set CE_PEER_NODE_ID to run a live two-node request/reply)");
  }

  console.log("RECIPE_OK 07_mesh_rpc");
}

main().catch((e) => {
  console.error(e);
  process.exitCode = 1;
});

File: recipes/07_mesh_rpc.rs
//! Recipe 07 — Mesh request/reply (the canonical app RPC pattern).
//!
//! Teaches: device-to-device app messaging over the mesh. A server subscribes to a topic and
//! answers inbound requests via `reply(reply_token, ...)`; a client calls `request(node, topic,
//! payload, timeout)` and blocks for the reply. This is the building block every CE *app* uses
//! for cross-node features (the node stays primitives-only).
//!
//! Run:  cargo run --example 07_mesh_rpc
//!
//! needs: two-nodes (a real request needs a peer running the reply loop). With one node we
//! exercise subscribe + the inbox snapshot, which is the single-node-testable slice of the
//! contract; the two-node request/reply is shown in comments + run when a peer is present.

use ce_rs::CeClient;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let base = std::env::var("CE_BASE_URL").unwrap_or_else(|_| ce_rs::DEFAULT_BASE_URL.into());
    let ce = CeClient::new(base);

    let topic = "cookbook/ping";

    // Server side: subscribe so inbound requests land in the inbox + stream.
    ce.subscribe(topic).await?;
    println!("subscribed to '{topic}'");

    // The reply loop an app would run (drains requests, answers each):
    //
    //   let stream = ce.messages_stream().await?;
    //   futures_util::pin_mut!(stream);
    //   while let Some(Ok(m)) = stream.next().await {
    //       if m.topic == topic {
    //           if let Some(tok) = m.reply_token {
    //               ce.reply(tok, b"pong").await?;
    //           }
    //       }
    //   }
    //
    // Client side, on a *different* node:
    //
    //   let reply = ce.request(server_node_id, topic, b"", 5000).await?;
    //   assert_eq!(reply, b"pong");

    // Single-node-testable slice: the inbox snapshot is reachable and returns a (possibly empty) list.
    let inbox = ce.messages().await?;
    println!("inbox snapshot: {} message(s)", inbox.len());

    // If a peer node id is supplied, perform a real round-trip.
    if let Ok(peer) = std::env::var("CE_PEER_NODE_ID") {
        match ce.request(&peer, topic, b"", 5000).await {
            Ok(reply) => println!("round-trip reply: {}", String::from_utf8_lossy(&reply)),
            Err(e) => println!("peer round-trip skipped: {e}"),
        }
    } else {
        println!("(set CE_PEER_NODE_ID to run a live two-node request/reply)");
    }

    println!("RECIPE_OK 07_mesh_rpc");
    Ok(())
}
